### PR TITLE
WIP e2e disk encryption tests

### PR DESCRIPTION
**Pull request by rolandmkunkel**
### Which issue this PR addresses:

very raw work in progress. no need to review yet. changes in predeploy json files will be moved to code

Fixes

### CODE REVIEW

File: deploy/generate.go [hunk 1]
```diff
@@ -2,6 +2,8 @@
 //go:generate go run ../vendor/github.com/go-bindata/go-bindata/go-bindata -nometadata -pkg $GOPACKAGE -ignore=generate.go -ignore=config.yaml -o ../pkg/deploy/bindata.go .
 //go:generate gofmt -s -l -w ../pkg/deploy/bindata.go
 
+// this first writes the bindata to the json, then the json to the bindata?
```
Comment (mjudeikis) [inline deploy/generate.go:5]: ?

File: pkg/deploy/generator/const.go [hunk 2]
```diff
@@ -15,7 +15,7 @@ const (
 	FileRPProductionParameters           = "rp-production-parameters.json"
 	FileRPProductionSubscription         = "rp-production-subscription.json"
 
-	FileClusterPredeploy       = "cluster-predeploy.json"
+	FileClusterPredeploy       = "cluster-predeploy.json" //maybe rename to e2e or development predeploy (and check if only used by e2e and hack scripts)
```
Comment (mjudeikis) [inline pkg/deploy/generator/const.go:18]: cluster-development-predeploy?

File: pkg/deploy/generator/resources_disk_encryption.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
+package generator
+
```
Comment (mjudeikis) [inline pkg/deploy/generator/resources_disk_encryption.go:2]: Missing header notes (autogenerated on `make generate`)

File: pkg/deploy/generator/resources_disk_encryption.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import (
+	mgmtcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-01/compute"
+	mgmtkeyvault "github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2019-09-01/keyvault"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+)
+
+func (g *generator) diskEncryptionKeyVault() *arm.Resource {
```
Comment (mjudeikis) [inline pkg/deploy/generator/resources_disk_encryption.go:12]: This is very similar to above keyvault function. Is there way to write one generic function to return keyvault and inject rules? Can be smaller re-usable one (See how RBAC is being generated)

File: pkg/deploy/generator/resources_disk_encryption.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		APIVersion: azureclient.APIVersion("Microsoft.KeyVault"),
+		// Only create a keyvault in CI. In local development we use a shared vault.
+		Condition: "[parameters('ci')]",
+	}
+}
+
+func (g *generator) diskEncryptionKey() *arm.Resource {
+	key := &mgmtkeyvault.Key{
+		KeyProperties: &mgmtkeyvault.KeyProperties{
+			Kty:     mgmtkeyvault.RSA,
```
Comment (mjudeikis) [inline pkg/deploy/generator/resources_disk_encryption.go:50]: You say RSA here but name is DES?

File: pkg/util/cluster/cluster.go [hunk 6]
```diff
@@ -168,6 +181,21 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
 		return err
 	}
 
+	kvName := sharedKeyVaultName
+	if c.ci {
+		// name is limited to 24 characters, but must be globally unique, so we generate one, try if it is availablet
+		kvName := "kv-" + uuid.Must(uuid.NewV4()).String()[:21]
+		result, err := c.keyVaults.CheckNameAvailability(ctx, mgmtkeyvault.VaultCheckNameAvailabilityParameters{Name: &kvName, Type: to.StringPtr("Microsoft.KeyVault/vaults")})
```
Comment (mjudeikis) [inline pkg/util/cluster/cluster.go:188]: So if this is not available? Maybe try again and not fail?

File: pkg/util/cluster/cluster.go [hunk 7]
```diff
@@ -168,6 +181,21 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
...
+		if err != nil {
+			return err
+		}
+		if result.NameAvailable != nil && !*result.NameAvailable {
+			return fmt.Errorf("Could not generate unique key vault name: %v", result.Reason)
+		}
+	}
+
+	diskEncryptionSetName := vnetResourceGroup + "-des"
```
Comment (mjudeikis) [inline pkg/util/cluster/cluster.go:197]: is this for read DES key?

**Review by mjudeikis — CHANGES_REQUESTED** [review]
Few nits but getting close

File: pkg/util/azureclient/keyvault/vaults.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package keyvault
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	mgmtkeyvault "github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2019-09-01/keyvault"
```
Comment (m1kola) [inline pkg/util/azureclient/keyvault/vaults.go:9]: `pkg/util/azureclient/keyvault` is for wrapping `github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault` client.

I guess we should follow the existing pattern for `github.com/Azure/azure-sdk-for-go/services/SERVICE_NAME/mgmt/...` clients and put its wrapper into `github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/keyvault`?

File: pkg/util/cluster/cluster.go [hunk 9]
```diff
@@ -324,6 +360,21 @@ func (c *Cluster) Delete(ctx context.Context, vnetResourceGroup, clusterName str
...
+			err = c.diskEncryptionSets.DeleteAndWait(ctx, vnetResourceGroup, vnetResourceGroup+"-des")
+			if err != nil {
+				errs = append(errs, err)
+			}
+
+			c.log.Info("deleting key for disk encryption")
+			_, err := c.keysClient.DeleteKey(ctx, fmt.Sprintf("https://%s.vault.azure.net", sharedKeyVaultName), clusterName+"-des-key") // possibly missing: [concat(parameters('kvName'), '/',
+			if err != nil {
+				errs = append(errs, err)
+			}
```
Comment (m1kola) [inline pkg/util/cluster/cluster.go:375]: Maybe put int into `else` above (somewhere after `c.log.Info("deleting route table")`)?

File: pkg/util/instancemetadata/instancemetadata.go [hunk 10]
```diff
@@ -27,6 +27,7 @@ type instanceMetadata struct {
 	subscriptionID string
 	location       string
 	resourceGroup  string
+	keyVaultName   string
```
Comment (m1kola) [inline pkg/util/instancemetadata/instancemetadata.go:30]: Why do we need this?

File: pkg/util/cluster/cluster.go [hunk 11]
```diff
@@ -193,20 +223,26 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
 		return err
 	}
 
+	diskEncryptionSet, err := c.diskEncryptionSets.Get(ctx, vnetResourceGroup, diskEncryptionSetName)
```
Comment (m1kola) [inline pkg/util/cluster/cluster.go:226]: I think at this point we should probably rename `vnetResourceGroup` to something more generic as it now hold not only vnet, but key vault as well. I however don't have ideas apart from something stupid like "bring your own resource resource group" :)

File: pkg/util/cluster/cluster.go (hunk 7 above)
Comment (m1kola) [reply pkg/util/cluster/cluster.go:197]: @mjudeikis I don't understand what you mean here. Could you please clarify?

This is the disk encryption set (DES) name which is going to be used for cluster creation with disk encryption.

File: pkg/util/cluster/cluster.go (hunk 11 above)
Comment (m1kola) [reply pkg/util/cluster/cluster.go:226]: Also specificly for this line:

You do `c.diskEncryptionSets.Get(...)` here but you only use it to retreive `diskEncryptionSet.ID`. At the same time, few lines below you construct this ID (line 235):

```go
{"/subscriptions/" + c.env.SubscriptionID() + "/resourceGroups/" + vnetResourceGroup + "/providers/Microsoft.Compute/diskEncryptionSets/" + diskEncryptionSetName, rbac.RoleReader},
```

File: pkg/util/azureclient/mgmt/compute/disks_addons.go [hunk 12]
```diff
@@ -20,3 +23,21 @@ func (c *disksClient) DeleteAndWait(ctx context.Context, resourceGroupName strin
...
+		result = append(result, page.Values()...)
+
+		err = page.NextWithContext(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
```
Comment (m1kola) [inline pkg/util/azureclient/mgmt/compute/disks_addons.go:43]: It looks like you are doing checks by listing VMs, not disks directly. Does this mean that we can remove this?

File: test/e2e/disk_encryption.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+			Expect(*vm.SecurityProfile.EncryptionAtHost).To(Equal(true))
+		}
+
+	})
+})
+
+var _ = Describe("Disk encryption at rest should be enabled with customer managed key", func() {
+	BeforeEach(skipIfNotInDevelopmentEnv)
+
+	// We have to get the disks by VM, because getting all disks by resource group does not work
```
Comment (m1kola) [inline test/e2e/disk_encryption.go:46]: I think it is fine to check via VMs. But I'm wondering why you can't list disks by resoruce group?

**Review by m1kola — COMMENTED** [review]
Few questions/suggestions, but in general looks good so far.

File: deploy/cluster-development-predeploy [hunk 14]
```diff
@@ -0,0 +1,199 @@
+{
```
Comment (rolandmkunkel) [inline deploy/cluster-development-predeploy:1]: renamed version of deploy/cluster-predeploy.json

**Review by m1kola — COMMENTED** [review]
I didn't finish reviewing this PR. Just positng early feedback.

**Review by m1kola — COMMENTED** [review]
Still need to review `pkg/util/cluster/cluster.go`.

File: pkg/deploy/generator/resources_disk_encryption.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+				{
+					TenantID: &tenantUUIDHack,
+					ObjectID: to.StringPtr("[parameters('rpServicePrincipalId')]"),
+					Permissions: &mgmtkeyvault.Permissions{
+						Keys: &[]mgmtkeyvault.KeyPermissions{
+							mgmtkeyvault.KeyPermissions(mgmtkeyvault.Create),
+							mgmtkeyvault.KeyPermissions(mgmtkeyvault.Delete),
+						},
+					},
+				},
```
Comment (m1kola) [inline pkg/deploy/generator/resources_disk_encryption.go:102]: Giving these permissions to RP service principal doesn't sound right. RP SP is for RP resources, not cluster/customer resoruces.

File: test/e2e/disk_encryption.go [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+		Expect(vms).NotTo(HaveLen(0))
+
+		By("checking the encryption property on each OS disk of each VM")
+		for _, vm := range vms {
+			osDisk, err := clients.Disks.Get(ctx, clusterResourceGroup, *vm.StorageProfile.OsDisk.Name)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(osDisk.Encryption.Type).To(Equal(mgmtcompute.EncryptionAtRestWithCustomerKey))
+		}
+
+		By("checking the encryption property on each data disk of each VM")
```
Comment (m1kola) [inline test/e2e/disk_encryption.go:90]: This should be about validating that default storrage class creates disks with disk encryption enabled.

See "How to test that default PV is encrypted" in for manual steps: https://github.com/Azure/ARO-RP/pull/1627

But I think this PR is too big already: it is very hard to follow. Let's deal wtih it once we merge this.
